chanzuckerberg/single-cell-data-portal
# feat: cxg conversion script updates for uns["spatial"]

## Reason for Change

## Changes

- added `convert_uns_to_cxg_group()`  handler for all uns["spatial"] actions as an additional step in `make_cxg()` script
- created `SpatialDataProcessor` class to handle `cxg` conversion for `uns` and  spatial deep zoom asset generation and upload
- added unit test coverage

## Testing steps

- using a jupyter notebook - run `make_cxg()` against a local `h5ad` - verify generated assets are uploaded to s3 and that the generated cxg includes `uns` group.
- verified with different use cases with:
-- datasets with fullres spatial image
-- datasets with hires only spatial image
-- datasets without spatial embedding

## Notes for Reviewer
- the package `libvips` had to be added to backed and processing Docker in order to get `pyvips` library (that generates the deep zoom tiles) to work properly
- the need to store uns["spatial"] in the cxg is to support the mvp implementation of spatial in explorer, where that image is generated on the fly - once deepzoom is launched - the step can be removed from the cxg conversion script
